pkg/dyninst/decode/types.go

```diff
@@ -0,0 +1,955 @@
+		return err
+	}
+	if notCaptured {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonPruned,
+		)
+	}
+	return nil
+}
+
+func (s *goSliceHeaderType) irType() ir.Type { return (*ir.GoSliceHeaderType)(s) }
+func (s *goSliceHeaderType) encodeValueFields(
+	d *Decoder,
+	enc *jsontext.Encoder,
+	data []byte) error {
+
+	if len(data) < int(s.ByteSize) {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonPruned,
+		)
+	}
+	if len(data) < 16 {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonPruned,
+		)
+	}
+	address := binary.NativeEndian.Uint64(data[0:8])
+	if address == 0 {
+		return writeTokens(enc,
+			jsontext.String("isNull"),
+			jsontext.Bool(true),
+		)
+	}
+	length := binary.NativeEndian.Uint64(data[8:16])
+	if length == 0 {
+		return writeTokens(enc,
+			jsontext.String("elements"),
+			jsontext.BeginArray,
+			jsontext.EndArray)
+	}
+	if err := writeTokens(enc,
+		jsontext.String("size"),
+		jsontext.Uint(length)); err != nil {
+		return err
+	}
+
+	elementSize := int(s.Data.Element.GetByteSize())
+	taa := typeAndAddr{
+		addr:   address,
+		irType: uint32(s.Data.GetID()),
+	}
+	sliceDataItem, ok := d.dataItems[taa]
+	if !ok {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonPruned,
+		)
+	}
+	if err := writeTokens(enc,
+		jsontext.String("elements"),
+		jsontext.BeginArray); err != nil {
+		return err
+	}
+	sliceLength := int(sliceDataItem.Header().Length) / elementSize
+	sliceData := sliceDataItem.Data()
+	var notCaptured = false
+	for i := range int(sliceLength) {
+		elementData := sliceData[i*int(s.Data.Element.GetByteSize()) : (i+1)*int(s.Data.Element.GetByteSize())]
+		if err := d.encodeValue(enc,
+			s.Data.Element.GetID(),
+			elementData,
+			s.Data.Element.GetName(),
+		); err != nil {
+			log.Tracef("could not encode slice element: %v", err)
+			notCapturedReason = notCapturedReasonPruned
+			break
+		}
+	}
+	if err := writeTokens(enc, jsontext.EndArray); err != nil {
+		return err
+	}
+	if length > uint64(sliceLength) {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonCollectionSize,
+		)
+	} else if notCaptured {
+		return writeTokens(enc,
+			notCapturedReason,
+			notCapturedReasonPruned,
+		)
+	}
+	return nil
+}
+
+func (s *goSliceDataType) irType() ir.Type { return (*ir.GoSliceDataType)(s) }
+func (s *goSliceDataType) encodeValueFields(
+	_ *Decoder,
+	enc *jsontext.Encoder,
+	_ []byte,
+) error {
+	return writeTokens(enc,
+		notCapturedReason,
+		notCapturedReasonUnimplemented,
+	)
+}
+
+func (s *goStringHeaderType) irType() ir.Type { return s.GoStringHeaderType }
+func (s *goStringHeaderType) encodeValueFields(
+	d *Decoder,
+	enc *jsontext.Encoder,
+	data []byte,
+) error {
+	realLength := binary.NativeEndian.Uint64(data[s.lenFieldOffset : s.lenFieldOffset+uint32(s.lenFieldSize)])
+	if err := writeTokens(enc,
+		jsontext.String("size"),
+		jsontext.Uint(realLength)); err != nil {
+		return err
+	}
```

Comment:
This should only be written if the value is truncated. The spec says: 
> When values are capped, an additional field is added to the Snapshot to indicate there is more data. The field truncated indicates that data were truncated because too long:
"truncated": true.  The original length of the string is also indicated by the size field only if truncated